endpoints/openrtb2/amp_auction.go

```diff
@@ -0,0 +1,230 @@
+package openrtb2
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/julienschmidt/httprouter"
+	"github.com/mssola/user_agent"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/exchange"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/pbs"
+	"github.com/prebid/prebid-server/pbsmetrics"
+	"github.com/prebid/prebid-server/stored_requests"
+	"net/http"
+	"strings"
+	"time"
+)
+
+type AmpResponse struct {
+	Targeting map[string]string `json:"targeting"`
+}
+
+// We need to modify the OpenRTB endpoint to handle AMP requests. This will basically modify the parsing
+// of the request, and the return value, using the OpenRTB machinery to handle everything inbetween.
+func NewAmpEndpoint(ex exchange.Exchange, validator openrtb_ext.BidderParamValidator, requestsById stored_requests.Fetcher, cfg *config.Configuration, met *pbsmetrics.Metrics) (httprouter.Handle, error) {
+	if ex == nil || validator == nil || requestsById == nil || cfg == nil || met == nil {
+		return nil, errors.New("NewEndpoint requires non-nil arguments.")
+	}
+
+	return httprouter.Handle((&endpointDeps{ex, validator, requestsById, cfg, met}).AmpAuction), nil
+}
+
+func (deps *endpointDeps) AmpAuction(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	// Prebid Server interprets request.tmax to be the maximum amount of time that a caller is willing
+	// to wait for bids. However, tmax may be defined in the Stored Request data.
+	//
+	// If so, then the trip to the backend might use a significant amount of this time.
+	// We can respect timeouts more accurately if we note the *real* start time, and use it
+	// to compute the auction timeout.
+	start := time.Now()
+
+	deps.metrics.RequestMeter.Mark(1)
+	// Set this as an AMP request in Metrics. Should it also count as an OpenRTB request?
+	deps.metrics.AmpRequestMeter.Mark(1)
+
+	req, errL := deps.parseAmpRequest(r)
+	isSafari := false
+	if ua := user_agent.New(r.Header.Get("User-Agent")); ua != nil {
+		name, _ := ua.Browser()
+		if name == "Safari" {
+			isSafari = true
+			deps.metrics.SafariRequestMeter.Mark(1)
+		}
+	}
+
+	if len(errL) > 0 {
```

Comment:
This block could be a `func writeError(errs []error, errMeter, go-metrics.Meter, w http.ResponseWriter) bool` to avoid duplicate code. The bool return is for whether or not it wrote an error, so the caller knows whether it needs to `return`